srv/salt/_modules/rgw.py

```diff
@@ -40,42 +40,37 @@ def configuration(role):
         role = 'rgw'
     if 'roles' in __pillar__:
         if 'rgw_configurations' in __pillar__:
-            for rgw_config in  __pillar__['rgw_configurations'].keys():
+            for rgw_config in  __pillar__['rgw_configurations']:
                 if rgw_config in role:
                     return rgw_config
-    return 
+    return
 
 
 
-def users(role):
+def users(realm='default'):
     """
-    Return the list of users.  Consider the default rgw and ganesha roles
-    equivalent.
+    Return the list of users for a realm.
     """
-    if 'roles' in __pillar__:
-        if 'rgw_configurations' in __pillar__:
-            if role == 'ganesha' or role == 'rgw':
-                # Special case for default names
-                users = [ u['uid'] for u in __pillar__['rgw_configurations']['rgw']['users'] ]
-                log.info("users: {}".format(users))
-                return users
-            if role in __pillar__['rgw_configurations']:
-                users = [ u['uid'] for u in __pillar__['rgw_configurations'][role]['users'] ]
-                log.info("users: {}".format(users))
-                return users
-        if 'rgw' in __pillar__['roles']:
-            return []
+    cmd = "radosgw-admin user list --rgw-realm={}".format(realm)
+    log.info("cmd: {}".format(cmd))
+    proc = Popen(cmd.split(), stdout=PIPE, stderr=PIPE)
+    proc.wait()
+    log.debug("rc: {}".format(proc.returncode))
+    if proc.returncode != '0':
+        return json.loads(proc.stdout.read())
     return []
 
-def add_users(pathname="/srv/salt/ceph/rgw/cache"):
+def add_users(pathname="/srv/salt/ceph/rgw/cache", jinja="/srv/salt/ceph/rgw/files/users.j2"):
     """
-    Write each user to its own file
+    Write each user to its own file.
     """
-    if 'rgw_configurations' not in __pillar__:
+    users = __salt__['slsutil.renderer'](jinja)
```

Comment:
we need to be careful with this because this call will only succeed if this function is called in the master node minion. Otherwise slsutil.renderer will not find the file `/srv/salt/ceph/rgw/files/users.j2`

Reply:
Many of our modules are in this category.  Much of it has to do with permissions as root or having a proper functioning return since orchestrations ignore runner return codes.